File: ios/DrinkScanAI/Modules/DrinkClassifierModule.m
/**
 * DrinkClassifierModule.m
 * Pure Objective-C CoreML classifier — no Swift, no bridging header needed.
 */

#import <React/RCTBridgeModule.h>
#import <CoreML/CoreML.h>
#import <Vision/Vision.h>
#import <UIKit/UIKit.h>

@interface DrinkClassifierModule : NSObject <RCTBridgeModule>
@property (nonatomic, strong) VNCoreMLModel *vnModel;
@property (nonatomic, strong) NSArray<NSString *> *classNames;
@property (nonatomic, assign) BOOL isLoaded;
@property (nonatomic, copy)   NSString *loadError;
@end

@implementation DrinkClassifierModule

RCT_EXPORT_MODULE(DrinkClassifierModule)

+ (BOOL)requiresMainQueueSetup { return NO; }

- (void)loadModelIfNeeded {
  if (self.isLoaded) return;

  // Find model in bundle
  NSURL *modelURL = [[NSBundle mainBundle] URLForResource:@"DrinkClassifier" withExtension:@"mlmodelc"]
                 ?: [[NSBundle mainBundle] URLForResource:@"DrinkClassifier" withExtension:@"mlpackage"];

  if (!modelURL) {
    // List bundle contents for debugging
    NSString *resourcePath = [[NSBundle mainBundle] resourcePath];
    NSArray *files = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:resourcePath error:nil];
    NSPredicate *pred = [NSPredicate predicateWithFormat:@"self CONTAINS[c] 'ml' OR self CONTAINS[c] 'Drink'"];
    NSArray *mlFiles = [files filteredArrayUsingPredicate:pred];
    self.loadError = [NSString stringWithFormat:@"DrinkClassifier not found. ML files in bundle: %@", mlFiles];
    NSLog(@"[DrinkClassifier] %@", self.loadError);
    return;
  }

  NSLog(@"[DrinkClassifier] Found model: %@", modelURL.path);

  NSError *error;
  MLModelConfiguration *config = [[MLModelConfiguration alloc] init];
  config.computeUnits = MLComputeUnitsAll;

  MLModel *mlModel = [MLModel modelWithContentsOfURL:modelURL configuration:config error:&error];
  if (!mlModel) {
    self.loadError = [NSString stringWithFormat:@"MLModel init failed: %@", error.localizedDescription];
    NSLog(@"[DrinkClassifier] %@", self.loadError);
    return;
  }

  self.vnModel = [VNCoreMLModel modelForMLModel:mlModel error:&error];
  if (!self.vnModel) {
    self.loadError = [NSString stringWithFormat:@"VNCoreMLModel failed: %@", error.localizedDescription];
    NSLog(@"[DrinkClassifier] %@", self.loadError);
    return;
  }

  // Load class names from drink_classes.json
  NSURL *jsonURL = [[NSBundle mainBundle] URLForResource:@"drink_classes" withExtension:@"json"];
  if (jsonURL) {
    NSData *data = [NSData dataWithContentsOfURL:jsonURL];
    NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
    NSArray *classes = json[@"classes"];
    NSArray *sorted = [classes sortedArrayUsingComparator:^NSComparisonResult(NSDictionary *a, NSDictionary *b) {
      return [a[@"index"] compare:b[@"index"]];
    }];
    NSMutableArray *names = [NSMutableArray array];
    for (NSDictionary *c in sorted) {
      [names addObject:c[@"name"] ?: @"unknown"];
    }
    self.classNames = [names copy];
    NSLog(@"[DrinkClassifier] Loaded %lu class names", (unsigned long)self.classNames.count);
  }

  self.isLoaded = YES;
  NSLog(@"[DrinkClassifier] Model loaded with %lu classes", (unsigned long)self.classNames.count);
}

RCT_EXPORT_METHOD(preloadModel:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
    [self loadModelIfNeeded];
    if (self.isLoaded) {
      resolve(@{@"loaded": @YES, @"classes": @(self.classNames.count)});
    } else {
      resolve(@{@"loaded": @NO, @"error": self.loadError ?: @"unknown error"});
    }
  });
}

RCT_EXPORT_METHOD(classifyImage:(NSString *)imagePath
                  topK:(nonnull NSNumber *)topK
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_USER_INITIATED, 0), ^{
    [self loadModelIfNeeded];

    if (!self.isLoaded) {
      reject(@"MODEL_ERROR", self.loadError ?: @"Model not loaded", nil);
      return;
    }

    UIImage *image = [UIImage imageWithContentsOfFile:imagePath];
    if (!image || !image.CGImage) {
      reject(@"IMAGE_ERROR", [NSString stringWithFormat:@"Cannot load image: %@", imagePath], nil);
      return;
    }

    __block NSMutableArray *output = [NSMutableArray array];
    dispatch_semaphore_t sem = dispatch_semaphore_create(0);

    VNCoreMLRequest *request = [[VNCoreMLRequest alloc] initWithModel:self.vnModel completionHandler:^(VNRequest *req, NSError *err) {
      NSArray<VNCoreMLFeatureValueObservation *> *results = req.results;
      VNCoreMLFeatureValueObservation *obs = results.firstObject;
      MLMultiArray *arr = obs.featureValue.multiArrayValue;
      if (!arr) { dispatch_semaphore_signal(sem); return; }

      NSInteger count = arr.count;
      NSMutableArray *logits = [NSMutableArray arrayWithCapacity:count];
      float maxL = -INFINITY;
      for (NSInteger i = 0; i < count; i++) {
        float v = arr[i].floatValue;
        [logits addObject:@(v)];
        if (v > maxL) maxL = v;
      }

      // Softmax
      float sum = 0;
      NSMutableArray *exps = [NSMutableArray arrayWithCapacity:count];
      for (NSNumber *l in logits) {
        float e = expf(l.floatValue - maxL);
        [exps addObject:@(e)];
        sum += e;
      }

      NSMutableArray *indexed = [NSMutableArray arrayWithCapacity:count];
      for (NSInteger i = 0; i < count; i++) {
        float prob = [exps[i] floatValue] / sum;
        [indexed addObject:@[@(i), @(prob)]];
      }

      [indexed sortUsingComparator:^NSComparisonResult(NSArray *a, NSArray *b) {
        return [b[1] compare:a[1]];
      }];

      NSInteger k = MIN(topK.integerValue, indexed.count);
      for (NSInteger i = 0; i < k; i++) {
        NSArray *item = indexed[i];
        NSInteger idx = [item[0] integerValue];
        float prob = [item[1] floatValue];
        NSString *name = (idx < (NSInteger)self.classNames.count) ? self.classNames[idx] : [NSString stringWithFormat:@"unknown_%ld", (long)idx];
        [output addObject:@{@"classIndex": @(idx), @"className": name, @"probability": @(prob)}];
      }
      dispatch_semaphore_signal(sem);
    }];

    request.imageCropAndScaleOption = VNImageCropAndScaleOptionCenterCrop;
    VNImageRequestHandler *handler = [[VNImageRequestHandler alloc] initWithCGImage:image.CGImage options:@{}];
    [handler performRequests:@[request] error:nil];
    dispatch_semaphore_wait(sem, DISPATCH_TIME_FOREVER);
    resolve(output);
  });
}

@end

File: ios/DrinkScanAI/Modules/DrinkClassifierModule.swift
import Foundation
import CoreML
import Vision
import UIKit

// Self-contained React Native module — no bridging header needed
// Uses @objc and NSObject directly

@objc(DrinkClassifierModule)
class DrinkClassifierModule: NSObject {

  private var vnModel: VNCoreMLModel?
  private var classNames: [String] = []
  private var isLoaded = false
  private var loadError: String?

  @objc static func requiresMainQueueSetup() -> Bool { return false }

  // ── Load model ───────────────────────────────────────────────────────────

  private func loadModel() {
    guard !isLoaded else { return }

    // Try .mlmodelc first (compiled), then .mlpackage
    let modelName = "DrinkClassifier"
    guard let modelURL = Bundle.main.url(forResource: modelName, withExtension: "mlmodelc")
                      ?? Bundle.main.url(forResource: modelName, withExtension: "mlpackage") else {
      loadError = "Model file '\(modelName).mlmodelc' not found in bundle. Files in bundle: \(bundleMLFiles())"
      NSLog("[DrinkClassifier] ERROR: %@", loadError!)
      return
    }

    NSLog("[DrinkClassifier] Found model at: %@", modelURL.path)

    do {
      let config = MLModelConfiguration()
      config.computeUnits = .all
      let mlModel = try MLModel(contentsOf: modelURL, configuration: config)
      vnModel = try VNCoreMLModel(for: mlModel)

      // Load class names from metadata
      if let classesJSON = mlModel.modelDescription.metadata[MLModelMetadataKey(rawValue: "classes")] as? String,
         let data = classesJSON.data(using: .utf8),
         let arr = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] {
        classNames = arr.sorted { ($0["index"] as? Int ?? 0) < ($1["index"] as? Int ?? 0) }
                       .compactMap { $0["name"] as? String }
        NSLog("[DrinkClassifier] Loaded %d class names from metadata", classNames.count)
      }

      // Fallback: load from drink_classes.json
      if classNames.isEmpty,
         let jsonURL = Bundle.main.url(forResource: "drink_classes", withExtension: "json"),
         let data = try? Data(contentsOf: jsonURL),
         let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
         let classes = json["classes"] as? [[String: Any]] {
        classNames = classes.sorted { ($0["index"] as? Int ?? 0) < ($1["index"] as? Int ?? 0) }
                            .compactMap { $0["name"] as? String }
        NSLog("[DrinkClassifier] Loaded %d class names from drink_classes.json", classNames.count)
      }

      isLoaded = true
      NSLog("[DrinkClassifier] Model loaded successfully with %d classes", classNames.count)
    } catch {
      loadError = "Failed to load model: \(error.localizedDescription)"
      NSLog("[DrinkClassifier] ERROR: %@", loadError!)
    }
  }

  private func bundleMLFiles() -> String {
    let fm = FileManager.default
    guard let resourcePath = Bundle.main.resourcePath else { return "no resourcePath" }
    let files = (try? fm.contentsOfDirectory(atPath: resourcePath)) ?? []
    return files.filter { $0.contains("ml") || $0.contains("ML") || $0.contains("Drink") }.joined(separator: ", ")
  }

  // ── Preload ───────────────────────────────────────────────────────────────

  @objc func preloadModel(_ resolve: @escaping RCTPromiseResolveBlock,
                           rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.global(qos: .background).async {
      self.loadModel()
      if self.isLoaded {
        resolve(["loaded": true, "classes": self.classNames.count])
      } else {
        resolve(["loaded": false, "error": self.loadError ?? "unknown"])
      }
    }
  }

  // ── Classify ──────────────────────────────────────────────────────────────

  @objc func classifyImage(_ imagePath: String,
                            topK: NSNumber,
                            resolver resolve: @escaping RCTPromiseResolveBlock,
                            rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.global(qos: .userInitiated).async {
      self.loadModel()

      guard self.isLoaded, let vnModel = self.vnModel else {
        reject("MODEL_ERROR", self.loadError ?? "Model not loaded", nil)
        return
      }

      guard let image = UIImage(contentsOfFile: imagePath),
            let cgImage = image.cgImage else {
        reject("IMAGE_ERROR", "Cannot load image: \(imagePath)", nil)
        return
      }

      var output: [[String: Any]] = []
      let semaphore = DispatchSemaphore(value: 0)

      let request = VNCoreMLRequest(model: vnModel) { req, err in
        defer { semaphore.signal() }
        guard let obs = req.results as? [VNCoreMLFeatureValueObservation],
              let arr = obs.first?.featureValue.multiArrayValue else { return }

        var logits = [Float]()
        for i in 0..<arr.count { logits.append(arr[i].floatValue) }

        let maxL = logits.max() ?? 0
        let exps = logits.map { exp($0 - maxL) }
        let sum  = exps.reduce(0, +)
        let probs = exps.map { $0 / sum }

        let k = min(topK.intValue, probs.count)
        let sorted = probs.enumerated().sorted { $0.element > $1.element }.prefix(k)
        output = sorted.map { (idx, prob) in
          let name = idx < self.classNames.count ? self.classNames[idx] : "unknown_\(idx)"
          return ["classIndex": idx, "className": name, "probability": Double(prob)]
        }
      }

      request.imageCropAndScaleOption = .centerCrop
      try? VNImageRequestHandler(cgImage: cgImage, options: [:]).perform([request])
      semaphore.wait()
      resolve(output)
    }
  }
}
